builder/bflag.go

```diff
@@ -0,0 +1,156 @@
+package builder
+
+import (
+	"fmt"
+	"strings"
+)
+
+type FlagType int
+
+const (
+	boolType FlagType = iota
+	stringType
+)
+
+type BuilderFlags struct {
+	Args  []string // actual flags/args from cmd line
+	flags map[string]*Flag
+	used  map[string]*Flag
+	Err   error
+}
+
+type Flag struct {
+	bf      *BuilderFlags
+	name    string
+	daType  FlagType
+	okEmpty bool
+	Value   string
+}
+
+func NewBuilderFlags() *BuilderFlags {
+	return &BuilderFlags{
+		flags: make(map[string]*Flag),
+		used:  make(map[string]*Flag),
+	}
+}
+
+func (bf *BuilderFlags) AddBool(name string, def bool) *Flag {
+	flag := bf.addFlag(name, boolType)
+	if flag == nil {
+		return nil
+	}
+	if def {
+		flag.Value = "true"
+	} else {
+		flag.Value = "false"
+	}
+	return flag
+}
+
+func (bf *BuilderFlags) AddString(name string, def string) *Flag {
+	flag := bf.addFlag(name, stringType)
+	if flag == nil {
+		return nil
+	}
+	flag.Value = def
+	return flag
+}
+
+func (bf *BuilderFlags) addFlag(name string, daType FlagType) *Flag {
+	if _, ok := bf.flags[name]; ok {
+		bf.Err = fmt.Errorf("Duplicate flag defined: %s", name)
+		return nil
+	}
+
+	newFlag := &Flag{
+		bf:     bf,
+		name:   name,
+		daType: daType,
+	}
+	bf.flags[name] = newFlag
+
+	return newFlag
+}
+
+func (fl *Flag) IsUsed() bool {
+	if _, ok := fl.bf.used[fl.name]; ok {
+		return true
+	}
+	return false
+}
+
+func (fl *Flag) IsTrue() bool {
+	if fl.daType != boolType {
+		// Should never get here
+		panic(fmt.Errorf("Trying to use IsTrue on a non-boolean: %s", fl.name))
```

Comment:
yea well, cpuguy83 had me change that  :-)

I still prefer “daType” to “flagType” but oh well

> On Apr 30, 2015, at 3:52 PM, Erik Hollensbe <notifications@github.com> wrote:
> 
> In builder/bflag.go <https://github.com/docker/docker/pull/10775#discussion_r29463634>:
> 
> > +	bf.flags[name] = newFlag
> > +
> > +	return newFlag
> > +}
> > +
> > +func (fl *Flag) IsUsed() bool {
> > +	if _, ok := fl.bf.used[fl.name]; ok {
> > +		return true
> > +	}
> > +	return false
> > +}
> > +
> > +func (fl *Flag) IsTrue() bool {
> > +	if fl.daType != boolType {
> > +		// Should never get here
> > +		panic(fmt.Errorf("Trying to use IsTrue on a non-boolean: %s", fl.name))
> Nice use of hungarian notation in "daType", fwiw. :P
> … <x-msg://16/#>
> —
> Reply to this email directly or view it on GitHub.
> 
>  <https://github.com/docker/docker/pull/10775#discussion_r29237111> <https://github.com/docker/docker/pull/10775/files#r29237111>
> —
> Reply to this email directly or view it on GitHub <https://github.com/docker/docker/pull/10775/files#r29463634>.
>